File: withPy/src/1.py
import collections
import time
from typing import Deque

def time_solution(string:str) -> None:
    start = time.time()
    print(solution(string))
    end = time.time()

    print(f"{end - start:.5f} sec")

def solution(string):
    str = []

    for char in string:
        if char.isalnum():
            str.append(char.lower())

    while len(str) > 1:
        if str.pop(0) != str.pop():
            return False

    return True

def solution1(string:str) -> bool:
    str: Deque = collections.deque()

    for char in string:
        if char.isalnum():
            str.append(char.lower())

    while len(str) > 1:
        if str.popleft() != str.pop():
            return False

    return True

time_solution("A man, a plan, a canal: Panama")
time_solution("race a car")
print(solution1("A man, a plan, a canal: Panama"))

File: withPy/src/1.ts
function time_solution(string: string):void {
    const start = Date.now();

    console.log(solution(string));

    const end = Date.now();

    console.log(`${end - start} ms`);
}

function solution(string: string):boolean {
    let str : string[] = [];

    for (let i = 0; i < string.length; i++) {
        if (string[i].match(/^[0-9a-zA-Z]$/)) {
            str.push(string[i].toLowerCase());
        }
    }

    for (let i = 0; str.length > 1; i++) {
        if (str.pop() != str.shift()) {
            return false;
        }
    }

    return true;
}

time_solution("A man, a plan, a canal: Panama");
time_solution("race a car");

export default solution;
